pkg/tfvars/internal/cache/cache.go

```diff
@@ -0,0 +1,184 @@
+			}
+		} else {
+			return "", err
+		}
+	}
+
+	return cacheDir, nil
+}
+
+// cacheFile puts data in the cache
+func cacheFile(reader io.Reader, filePath string) (err error) {
+	logrus.Debugf("Unpacking file into %q...", filePath)
+
+	flockPath := fmt.Sprintf("%s.lock", filePath)
+	flock, err := os.Create(flockPath)
+	if err != nil {
+		return err
+	}
+	defer flock.Close()
+	defer func() {
+		err2 := os.Remove(flockPath)
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	err = unix.Flock(int(flock.Fd()), unix.LOCK_EX)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err2 := unix.Flock(int(flock.Fd()), unix.LOCK_UN)
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	_, err = os.Stat(filePath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil // another cacheFile beat us to it
+	}
+
+	tempPath := fmt.Sprintf("%s.tmp", filePath)
+	file, err := os.OpenFile(tempPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0444)
+	if err != nil {
+		return err
+	}
+	closed := false
+	defer func() {
+		if !closed {
+			file.Close()
+		}
+	}()
+
+	_, err = io.Copy(file, reader)
+	if err != nil {
+		return err
+	}
+
+	err = file.Close()
+	if err != nil {
+		return err
+	}
+	closed = true
+
+	return os.Rename(tempPath, filePath)
+}
+
+// DownloadFile obtains a file from a given URL, puts it in the cache folder, defined by dataType parameter,
+// and returns the local file path.
+// If the query string contains sha256 parameter (i.e. https://example.com/data.bin?sha256=098a5a...),
+// then the downloaded data checksum will be compared with the provided value.
+func DownloadFile(baseURL string, dataType string) (string, error) {
+	// Convert the given URL into a file name using md5 algorithm
+	fileName := fmt.Sprintf("%x", md5.Sum([]byte(baseURL)))
+
+	cacheDir, err := getCacheDir(dataType)
+	if err != nil {
+		return "", err
+	}
+	filePath := filepath.Join(cacheDir, fileName)
+
+	// If the file has already been cached, return its path
+	_, err = os.Stat(filePath)
+	if err == nil {
+		logrus.Infof("The file was found in cache: %v. Reusing...", filePath)
+		return filePath, nil
+	}
+	if !os.IsNotExist(err) {
+		return "", err
+	}
+
+	// Send a request
+	resp, err := http.Get(baseURL)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.Errorf("bad status: %s", resp.Status)
+	}
+
+	hasher := sha256.New()
+	data := io.TeeReader(resp.Body, hasher)
+
+	err = cacheFile(data, filePath)
+	if err != nil {
+		return "", err
+	}
+
+	// Validate sha256 checksum if it was provided as a part of the URL
+	parsedURL, err := url.ParseRequestURI(baseURL)
+	if err != nil {
+		return "", err
+	}
+	if sha256Checksums, ok := parsedURL.Query()["sha256"]; ok {
+		sha256Checksum := sha256Checksums[0]
+
+		foundChecksum := fmt.Sprintf("%x", hasher.Sum(nil))
+		if sha256Checksum != foundChecksum {
+			logrus.Error("File sha256 checksum is invalid.")
+			defer os.Remove(filePath)
```

Comment:
this is racey, because the cachedFile is already created on the disk, somebody else will use that file without integrity check...

Reply:
done! now we check the temporary file and if its checksum is invalid it won't be cached